File: citrussoda/4week/104_maximum-depth-of-binary-tree.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number}
 */
var maxDepth = function (root) {
  if (root === null) return 0;

  return dig(root);
};

function dig(node) {
  if (node === null) return 0;

  let leftDepth = dig(node.left);
  let rightDepth = dig(node.right);

  return Math.max(leftDepth, rightDepth) + 1;
}

File: citrussoda/4week/cpp/104_maximum-depth-of-binary-tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int maxDepth(TreeNode* root) {
        // nullptr이 NULL보다 빠르며 NULL은 정수 0으로 인식이 되나 nullptr이 인식되지 않음.
        if (root == nullptr) return 0;

        // 따로 변수를 선언하지 않고 파라미터 내부에서 직접 재귀 가능
        return max(maxDepth(root -> left) + 1, maxDepth(root -> right) + 1);
    }
};